packages/next/build/index.ts

```diff
@@ -1088,7 +1088,125 @@ export default async function build(
+          const routeRegex = getRouteRegex(dataRoute.replace(/\.json$/, ''))
+
+          dataRouteRegex = normalizeRouteRegex(
+            routeRegex.re.source.replace(/\(\?:\\\/\)\?\$$/, `\\.json$`)
+          )
+          namedDataRouteRegex = routeRegex.namedRegex!.replace(
+            /\(\?:\/\)\?\$$/,
+            `\\.json$`
+          )
+          routeKeys = routeRegex.routeKeys
+        } else {
+          dataRouteRegex = normalizeRouteRegex(
+            new RegExp(
+              `^${path.posix.join(
+                '/_next/data',
+                escapeStringRegexp(buildId),
+                `${pagePath}.json`
+              )}$`
+            ).source
+          )
+        }
+
+        return {
+          page,
+          routeKeys,
+          dataRouteRegex,
+          namedDataRouteRegex,
+        }
+      })
+
+      await promises.writeFile(
+        routesManifestPath,
+        JSON.stringify(routesManifest),
+        'utf8'
+      )
+    }
+
+    // Since custom _app.js can wrap the 404 page we have to opt-out of static optimization if it has getInitialProps
+    // Only export the static 404 when there is no /_error present
+    const useStatic404 =
+      !hasConcurrentFeatures &&
+      !customAppGetInitialProps &&
+      (!hasNonStaticErrorPage || hasPages404)
+
+    if (invalidPages.size > 0) {
+      const err = new Error(
+        `Build optimization failed: found page${
+          invalidPages.size === 1 ? '' : 's'
+        } without a React Component as default export in \n${[...invalidPages]
+          .map((pg) => `pages${pg}`)
+          .join(
+            '\n'
+          )}\n\nSee https://nextjs.org/docs/messages/page-without-valid-component for more info.\n`
+      ) as NextError
+      err.code = 'BUILD_OPTIMIZATION_FAILED'
+      throw err
+    }
+
+    await writeBuildId(distDir, buildId)
+
+    if (config.experimental.optimizeCss) {
+      const cssFilePaths = getCssFilePaths(buildManifest)
+
+      requiredServerFiles.files.push(
+        ...cssFilePaths.map((filePath) => path.join(config.distDir, filePath))
+      )
+    }
+
+    const features: EventBuildFeatureUsage[] = [
+      {
+        featureName: 'experimental/optimizeCss',
+        invocationCount: config.experimental.optimizeCss ? 1 : 0,
+      },
+      {
+        featureName: 'optimizeFonts',
+        invocationCount: config.optimizeFonts ? 1 : 0,
+      },
+    ]
+    telemetry.record(
+      features.map((feature) => {
+        return {
+          eventName: EVENT_BUILD_FEATURE_USAGE,
+          payload: feature,
+        }
+      })
+    )
+
+    await promises.writeFile(
+      path.join(distDir, SERVER_FILES_MANIFEST),
+      JSON.stringify(requiredServerFiles),
+      'utf8'
+    )
+
+    if (postCompileSpinner) postCompileSpinner.stopAndPersist()
+
     if (config.outputFileTracing) {
+      const outputFileTracingSpinner = createSpinner({
+        prefixText: `${Log.prefixes.info} Tracing output files`,
```

Comment:
This is only tracing `next-server`, the majority of output file tracing will be in the production build section since it's in the webpack plugin, do we want to make sure the spinner shows then instead?

Reply:
hmm... I guess we can't show nested spinners...

Reply:
Ah yeah might be tricky to manage that, should we rename this to `Tracing next-server` for now then?